docs/rfcs/20201208_fragmentation.md

```diff
@@ -0,0 +1,303 @@
+
+ - Is the huge allocator used unconditionally for allocations above some size, or only when the buddy allocator fails?
+ - Is the huge pool per-shard or shared?
+ - Do we just use the native built-in libc allocator (via the saved malloc pointer) as the huge poor, or layer another allocator on top of the reserved storage?
+ - How exactly does the allocator reserve physical memory, since they mostly work in the virtual address domain? We need an allocator which carefully limits its physical memory use, but tries to use all of it e.g., by punching holes (e.g., with `MADV_DONTNEED`) in existing mappings.
+
+
+### Abandon the Allocator
+
+We know seastar works with the default system allocator: that's how we run it in debug mode (so we can use ASAN and friends). So in principle, we could use the system allocator, or some enhanced drop-in replacement instead. However, we would lose the significant benefits of the seastar allocator:
+
+ - The performance benefits of per-shard allocation such as no locks or thread-local caches, exact NUMA affinity, etc.
+ - Allocation may block the reactor threads: the seastar allocator minimizes the chance of blocking my allocating everything up front, but default allocators calling mmap may block in some cases.
+ - The reclaim mechanism.
+ - Heap profiling (though the replacement may offer similar functionality).
+ - The exact division of memory among shards and the exact up-front nature of memory allocation. General purpose allocators tend to keep requesting more memory from the OS until something breaks, while the seastar allocator fails immediately if the assigned space is exhausted (after reclaim).
+
+This would be a significant and potentially risky change. 
+
+### Quality of Life
+
+Even though we might be living with fragmentation, we can make some quality of life improvements to make our life easier when root-causing failures.
+
+ - OOM error messages should clearly log relevant details such as the "availability" metric discussed above to pinpoint if the failure was caused by fragmentation.
+ - OOM error messages should include details of reclaim activity (if any) done during the failing allocation.
+ - We should surface fragmentation related metrics (perhaps availability at a few key sizes) in our metrics.
+ - We should surface metrics related to reclaim and batch cache activity to detect the "partial failure" case where the batch cache is thrashing described above (needs some thought about heuristics to detect this case).
+ - We should track the high water mark in the seastar allocator and expose it, so we can see how it evolves over time. 
```

Comment:
Yeah I think RSS would be an approximation for the sum of the process-wide allocator HWMs in production, tho at the SS level is a per-shard metric (since allocators are per-shard).

Yes, the batch cache means that we tend torwards full memory use as long as there is some activity and this is by design. Especially in production scenarios the RSS will probably not drop much even if activity slows down or stops as we don't expect competition for pages (especially since we try not to use the page cache).